cli/main.c

```diff
@@ -1317,8 +1317,12 @@ static int fw_update(int argc, char **argv)
 {
 	int ret;
 	int type;
-	const char *desc = "Flash the firmware with a new image";
-
+	const char *desc = "Flash device with a new firmware image\n\n"
+			   "NOTE - A device can be in one of these three boot phases: \n"
+			   "Bootloader1 (BL1), Bootloader2 (BL2) or MAIN FIRMWARE (FW).\n\n"
```

Comment:
Here is my understanding of the mfg submenu:
- Unlike main menu, whose commands are more or less intuitive to most users, the mfg submenu commands are used by advanced users during manufacturing process. Users are directed to these commands by a Knowledge-Base page (e.g., How to Recover Device from Corrupt Partitions, or How to Generate Secure Setting Files and Load to Device). These KB pages are available online and gives background on what BL1/2 are, how to boot to BL1/2, and the procedure and commands to use at each step.
- So the help information in mfg submenu should provide enough information to help user if he/she got stuck with a command in particular (e.g., missing parameters, wrong boot phase, wrong file format), but leave the background knowledge to the KB page.
- This is similar to GAS submenu, where we did not detail which register/bit controls what. User need to find those info elsewhere, before they can use the commands there. Of course GAS is a more extreme case where you basically cannot use the commands without background knowledge, while mfg submenu command is more in the middle: you can play with it but background knowledge is still required.

Reply:
No, that's a horrible idea. All commands should be largely understandable from the help text regardless of whether the user got send there from more detailed documentation. Saying you don't have to document something because it's documented somewhere non-obvious is wrong.

The analogy with the GAS sub-menu is flawed. It provides an interface to raw registers that no one should ever really use and is documented as such.

If you want to document the mfg submenu as more advanced commands, fine; but that does not mean you don't need proper documentation to say what the commands are used for and how to use this. This is especially so because we've made things more confusing with having a second way to send firmware to the device and thus the documentation must be *crystal* clear on the differences between them without assuming the user's read some random knowledge base article somewhere else.

Additionally, just because there will be documentation somewhere in some non-obvious knowledge- base doesn't mean making the procedure simpler with a fw-recover command would not be valuable.

Reply:
Agree on adding more detailed descriptions of boot phases. I have put the descriptions in common.h so they can be easily included in other commands if needed. Please take a look.

Renaming 'fw-transfer' to 'fw-transfer-bl2' is a good idea, will be applied when we check in the corresponding code. 

About having one commands that does transferring BL2, updating BL2/MAIN and all: there are many branches in recovery process: after transferring BL2, user can choose to switch to UART interface for future commands; also user have the option to update BL2/CFG/MAIN to active/non-active partition, and choose to boot from active/non-active partitions using CFG from active/non-active partition. Using one command means user have to set many command options for this command. It would be much easier if we offer them as individual commands instead, so user can do mix and match to suit their needs.

About keeping a copy of baseline BL2 image that works for most cases, as firmware and utility code are both constantly evolving, it is hard to find a version of BL2 that support most cases, for different hardware revisions. Over time, a baseline BL2 image will slowly become obsolete, if we don't update them along with the utility code.

@lsgunth let me know what you think.

Reply:
Well my comment for a fw-recover command would have been in addition to the fw-transfer command. Just a thought to make it easier on people.